game_mechanics.py

```diff
@@ -1,71 +1,449 @@
+            opponent_action = self._opponent_choose_move(-self._board)
+            opponent_reward = self._step(opponent_action, verbose)  # Can be None, fix
+            # Negative sign is because the opponent's victory is your loss
+            reward -= opponent_reward
+
+        if self.done:
+            if np.sum(self._board == 1) > np.sum(self._board == -1):
+                reward = 1
+            elif np.sum(self._board == 1) < np.sum(self._board == -1):
+                reward = -1
+            else:
+                reward = 0
+        else:
+            reward = 0
+
+        return self._board, reward, self.done, self.info
+
+    @property
+    def game_over(self) -> bool:
+        return (
+            not has_legal_move(self._board, self._player)
+            and not has_legal_move(self._board, self._player * -1)
+            or self.running_tile_count == self.board_dim**2
+        )
+
+    ## TODO: Pygame stuff
+    # def __del__(self):
+    #     """Destructor, quit pygame if game over."""
+    #     if self._screen is not None:
+    #         pygame.quit()
+
+    # def render(self) -> None:
+    #     """Renders game in pygame."""
+    #     if self._screen is None:
+    #         pygame.init()
+    #         self._screen = pygame.display.set_mode(
+    #             (self.SQUARE_SIZE * self.N_COLS, self.SQUARE_SIZE * self.N_ROWS)
+    #         )
+
+    #     # Draw background of the board
+    #     pygame.gfxdraw.box(
+    #         self._screen,
+    #         pygame.Rect(
+    #             0,
+    #             0,
+    #             self.N_COLS * self.SQUARE_SIZE,
+    #             self.N_ROWS * self.SQUARE_SIZE,
+    #         ),
+    #         self.BLUE_COLOR,
+    #     )
+
+    #     # Draw the circles - either as spaces if filled or
+    #     for r in range(self.N_ROWS):
+    #         for c in range(self.N_COLS):
+    #             space = self._board[r, c]
+    #             colour = (
+    #                 self.RED_COLOR
+    #                 if space == 1
+    #                 else self.YELLOW_COLOR
+    #                 if space == -1
+    #                 else self.BACKGROUND_COLOR
+    #             )
+
+    #             # Anti-aliased circle drawing
+    #             pygame.gfxdraw.aacircle(
+    #                 self._screen,
+    #                 c * self.SQUARE_SIZE + self.SQUARE_SIZE // 2,
+    #                 r * self.SQUARE_SIZE + self.SQUARE_SIZE // 2,
+    #                 int(self.DISC_SIZE_RATIO * self.SQUARE_SIZE / 2),
+    #                 colour,
+    #             )
+
+    #             pygame.gfxdraw.filled_circle(
+    #                 self._screen,
+    #                 c * self.SQUARE_SIZE + self.SQUARE_SIZE // 2,
+    #                 r * self.SQUARE_SIZE + self.SQUARE_SIZE // 2,
+    #                 int(self.DISC_SIZE_RATIO * self.SQUARE_SIZE / 2),
+    #                 colour,
+    #             )
+    #     pygame.display.update()
```

Comment:
TODO now or TODO later?

Side note: This VScode plugin may be useful for you to track TODO comments
https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree

Reply:
It's kinda important but no one complained at the lack of render option in last cohort

Reply:
Add a ticket?